### PR TITLE
#Issue-57 Controller generation

### PR DESCRIPTION
**Pull request by serhiyzhovnir**
### Description (*)
This PR adds Magento Controller generation action.
Screenshot 2020-04-22 at 08 41 12

Frontend class generation:
Screenshot 2020-04-22 at 08 41 59
Screenshot 2020-04-22 at 08 41 36

Backend class generation:
Screenshot 2020-04-22 at 08 44 22
Screenshot 2020-04-22 at 08 44 37

#### Additional Adjustments
The class name validation was added to the following validators:
-  OverrideClassByAPreferenceDialogValidator
-  NewViewModelValidator
-  NewGraphQlResolverValidator
-  NewCronjobValidator
-  NewBlockValidator
-  CreateAnObserverDialogValidator
-  CreateAPluginDialogValidator

### Fixed Issues (if relevant)

1. magento/magento2-phpstorm-plugin#57: Action/Code Generation. Controller generation

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages

### CODE REVIEW

File: resources/fileTemplates/code/Magento Controller Class Body.php.ft [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#if (${ACL})
```
Comment (VitaliyBoyko) [inline resources/fileTemplates/code/Magento Controller Class Body.php.ft:1]: Hi
I suggest calling this fille like `Magento Controller Adminhtml`

File: resources/fileTemplates/internal/Magento Module Controller Class.php.ft [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<?php
```
Comment (VitaliyBoyko) [inline resources/fileTemplates/internal/Magento Module Controller Class.php.ft:1]: And this `Magento Controller Frontend`

File: resources/META-INF/plugin.xml [hunk 3]
```diff
@@ -57,6 +57,7 @@
         <!-- Module file generators -->
         <group id="MagentoNewModuleFileGroup" class="com.magento.idea.magento2plugin.actions.groups.NewModuleFileGroup" text="Module File" popup="true">
             <action id="MagentoCreateABlock" class="com.magento.idea.magento2plugin.actions.generation.NewBlockAction" />
+            <action id="MagentoCreateAController" class="com.magento.idea.magento2plugin.actions.generation.NewControllerAction" />
             <action id="MagentoCreateACronjob" class="com.magento.idea.magento2plugin.actions.generation.NewCronjobAction" />
             <action id="MagentoCreateAViewModel" class="com.magento.idea.magento2plugin.actions.generation.NewViewModelAction" />
             <action id="MagentoCreateAGraphQlResolver" class="com.magento.idea.magento2plugin.actions.generation.NewGraphQlResolverAction" />
```
Comment (VitaliyBoyko) [inline resources/META-INF/plugin.xml:63]: Internal templates should be added to plugin.xml to be editable from the preferences.

File: src/com/magento/idea/magento2plugin/actions/generation/dialog/NewControllerDialog.form [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        <clientProperties>
+          <html.disable class="java.lang.Boolean" value="false"/>
+        </clientProperties>
+      </component>
+      <component id="ca20c" class="javax.swing.JCheckBox" binding="inheritClass">
+        <constraints>
+          <grid row="4" column="0" row-span="1" col-span="1" vsize-policy="0" hsize-policy="3" anchor="8" fill="0" indent="0" use-parent-layout="false"/>
+        </constraints>
+        <properties>
+          <text value="Inherit Action Class"/>
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/dialog/NewControllerDialog.form:159]: The bundle should be used for phrases.

File: src/com/magento/idea/magento2plugin/actions/generation/dialog/NewControllerDialog.form [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+              </grid>
+            </constraints>
+            <properties/>
+          </component>
+          <component id="4eaa9" class="javax.swing.JLabel">
+            <constraints>
+              <grid row="0" column="0" row-span="1" col-span="1" vsize-policy="0" hsize-policy="0" anchor="8" fill="0" indent="0" use-parent-layout="false"/>
+            </constraints>
+            <properties>
+              <text value="Admin Resource ACL"/>
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/dialog/NewControllerDialog.form:186]: Use the bundle.

File: src/com/magento/idea/magento2plugin/actions/generation/dialog/NewControllerDialog.java [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+import javax.swing.*;
+import java.awt.event.*;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+public class NewControllerDialog extends AbstractDialog {
+    public static String ADMINHTML_AREA = "adminhtml";
+    public static String FRONTEND_AREA = "frontend";
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/dialog/NewControllerDialog.java:27]: We already have `com.magento.idea.magento2plugin.magento.packages.Package.Areas#adminhtml` and `com.magento.idea.magento2plugin.magento.packages.Package.Areas#frontend`

File: src/com/magento/idea/magento2plugin/actions/generation/dialog/validator/NewControllerValidator.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    public NewControllerValidator()
+    {
+        this.validatorBundle = new ValidatorBundle();
+    }
+
+    public boolean validate()
+    {
+        String errorTitle = "Error";
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/dialog/validator/NewControllerValidator.java:35]: Use bundle.

Comment (VitaliyBoyko) [reply src/com/magento/idea/magento2plugin/actions/generation/dialog/validator/NewControllerValidator.java:35]: We about to move all the rest as well.

File: src/com/magento/idea/magento2plugin/actions/generation/generator/ModuleControllerClassGenerator.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+            }
+        }
+
+        attributes.setProperty("USES", formatUses(uses));
+    }
+
+    private ArrayList<String> getUses()
+    {
+        return new ArrayList<>(Arrays.asList(
+                "Magento\\Framework\\App\\ResponseInterface",
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/generator/ModuleControllerClassGenerator.java:180]: Let's make constants under Magento directory, see examples: `com.magento.idea.magento2plugin.magento.files.GraphQlResolver`

File: src/com/magento/idea/magento2plugin/magento/files/ControllerPhp.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+        return httpMethodInterfaces.get(method);
+    }
+
+    private static Map<String, String> getHttpMethodInterfaces()
+    {
+        if (httpMethodInterfaces == null) {
+            httpMethodInterfaces = new HashMap<String, String>();
+
+            httpMethodInterfaces.put("GET", "Magento\\Framework\\App\\Action\\HttpGetActionInterface");
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/magento/files/ControllerPhp.java:66]: Reimplement it using `enum` as I've mentioned before

File: src/com/magento/idea/magento2plugin/actions/generation/dialog/NewControllerDialog.java [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+    public static String HTTP_GET = "GET";
+    public static String HTTP_POST = "POST";
+    public static String HTTP_DELETE = "DELETE";
+    public static String HTTP_PUT = "PUT";
```
Comment (VitaliyBoyko) [inline src/com/magento/idea/magento2plugin/actions/generation/dialog/NewControllerDialog.java:31]: Create enum with all types of requests in the separate class under /magento/packages as it was made for areas. Thank you!

**Comment by serhiyzhovnir** [comment]
Hello @VitaliyBoyko 
The PR was adjusted according to the recommendations.
Could you, please, review it again?
Thank you!

**Comment by eduard13** [comment]
@VitaliyBoyko please keep in mind the changes from #227 while reviewing this one.
Thank you.

**Comment by serhiyzhovnir** [comment]
Hi @eduard13 @VitaliyBoyko 
The PR was adjusted according to the changes that were added in https://github.com/magento/magento2-phpstorm-plugin/pull/227 PR.
Thank you!

**Comment by serhiyzhovnir** [comment]
Hi @eduard13 
The requested changes were added and the merge conflict is resolved.
Please, check the PR again.
Thank you!

File: resources/fileTemplates/internal/Magento Module Controller Backend Class.php.ft [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+     * Authorization level of a basic admin session
+     */
+    const ADMIN_RESOURCE = '${ACL}';
+
+#end
+    /**
+     * Execute action based on request and return result
+     *
+     * @return ResultInterface|ResponseInterface
+     * @throws NotFoundException
```
Comment (eduard13) [inline resources/fileTemplates/internal/Magento Module Controller Backend Class.php.ft:25]: Is there any reason of throwing `NotFoundException ` at the moment? 🤔

Comment (serhiyzhovnir) [reply resources/fileTemplates/internal/Magento Module Controller Backend Class.php.ft:25]: Yes, all controllers implement [ActionInterface](https://github.com/magento/magento2/blob/2.3/lib/internal/Magento/Framework/App/ActionInterface.php) and the execute method of action interface throws `NotFoundException`.
Code reference: https://github.com/magento/magento2/blob/2.3/lib/internal/Magento/Framework/App/ActionInterface.php#L32

**Pull request merged**